File: state/cpp/state.cc
#include "state/cpp/state.h"

void Context::Next() {
    State *old_state = state;
    old_state->Handle(this);
    delete old_state;
}

Context::~Context() {
    if(state != nullptr) {
        delete state;
    }
}

void DayState::Handle(Context *ctx) {
    const std::string days[7] = {"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"};
    for(int i=0; i<7; ++i) {
        if(days[i] == day) {
            ctx->Set(new DayState(days[(i+1)%7]));
        }
    }
}

std::string DayState::Get() const {
    return day;
}

void Context::Set(State *state_) {
    state = state_;
}

DayContext::DayContext() {
    state = new DayState("Sunday");
}

std::string DayContext::Today() const {
    if(state == nullptr) {
        return "Can not get today";
    }
    return state->Get();
}

File: state/state.go
package state

type Context interface {
	Next()
}

type State interface {
	Handle(ctx Context)
}

type DayContext struct {
	state DayState
}

type DayState struct {
	day string
}

func (d DayState) Get() string {
	return d.day
}

func newDayState(day string) DayState {
	return DayState{
		day: day,
	}
}

func (d *DayContext) Next() {
	d.state.Handle(d)
}

func (d *DayContext) Today() string {
	return d.state.Get()
}

func (d DayState) Handle(ctx *DayContext) {
	days := []string{"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"}
	for i, day := range days {
		if day == d.Get() {
			ctx.state = newDayState(days[(i+1)%7])
		}
	}
}

func NewDayContext() *DayContext {
	return &DayContext{
		state: newDayState("Sunday"),
	}
}
